### PR TITLE
Implements the employee /location workflow 

### LINKED ISSUE #7: Issue-7-[EMP-2] Update Work Location Command
---
name: Task
about:  Update Work Location Command
labels: task, feature, employee
---

## Summary

Implement `/location` slash command allowing employees to update their work location (Office/WFH) for a selected date with validation and confirmation.

## Task Details

- [ ] Implement `/location` command handler
- [ ] Add command to Discord registration with parameters: `date`, `location`
- [ ] Parse interaction options from webhook payload
- [ ] Validate date format (YYYY-MM-DD)
- [ ] Validate date is not in past
- [ ] Validate date is within 7 days from today
- [ ] Enforce cut-off rule: no updates after previous day 9 PM (Asia/Dhaka timezone)
- [ ] Create or update location in DynamoDB user document
- [ ] Return Discord interaction response with confirmation
- [ ] Handle edge cases (invalid date, invalid location)
- [ ] Add error messages for validation failures
- [ ] Write unit tests for command handler
- [ ] Test command in Discord test server

## Context

This is **Issue 7** (EMP-2) - second employee feature. This enables employees to indicate whether they'll be in office or working from home.

**Business Rules:**
- Date must be today or future (no past dates)
- Date must be within 7 days from today (matches meal command window)
- Cut-off rule: updates for date D blocked after D-1 9:00 PM (Asia/Dhaka timezone)
- Valid locations: office, wfh
- User must exist

## Resources

- [discord.py Command Documentation](https://discordpy.readthedocs.io/en/stable/interactions/api.html)
- [tech-spec.md](tech-spec.md)

### PR DESCRIPTION
**Pull request by najmul-craftsmen**
Closes #7 

## Changelog

**Feature:** Employees can now update their work location (Office/WFH) for any date within the next 7 days using the `/location` command. Responses include clear validation messages and enforce cut-off times to prevent schedule conflicts.

## What does this PR do?

Implements the employee `/location` workflow end-to-end on the Discord bot, including command registration, interaction handling, DynamoDB persistence, validation rules, and cut-off enforcement.

## Type of Change

- [X] Feature (new employee command)
- [X] Behavior change (validation and cut-off rules)
- [X] Refactor (modular handler structure for maintainability)

### 1. Discord Slash Command Registration

Updated `register_commands.py`:

- Added `/location` command with options:
	- `location` (office, wfh) - required, with emoji choices
	- `date` (required, supports `today` or `YYYY-MM-DD`)
- Consistent with `/meal` command structure for unified UX

### 2. Location Command Handler Module

Created new `src/handlers/location_handler.py` for code maintainability:

- `handle_location_command(...)` - main handler function
- Parses location and date from Discord interaction payload
- Date parsing supports `today` shortcut + `YYYY-MM-DD` format
- Location enum conversion (`WorkLocationType`)
- DynamoDB read/write integration via `DynamoDBRepository`
- Helper functions for date validation and response formatting

### 3. Command Routing Integration

Updated `src/handlers/discord_handler.py`:

- Added command routing for `location` in `handle_application_command`
- Import `handle_location_command` from new location handler module
- Maintains clean separation of concerns

### 4. Validation and Business Rules Implemented

- Date must not be in the past (bot-local date, Asia/Dhaka timezone)
- Date must be within next 7 days (aligned with meal command)
- Cut-off rule enforced:
	- for target date `D`, updates close at `D-1 9:00 PM`
	- timezone: `Asia/Dhaka`
- User must exist (no auto-creation; created by `/meal` command)
- Invalid input returns clear ephemeral error responses

### 5. Response Message Format

- Simplified location-only feedback (no meal status clutter)
- Shows confirmation + location emoji and label
- Ephemeral message (visible only to user)

**Response Example:**
```
✅ **Office** set for 2026-03-15
🏢 Location: OFFICE
```

## Files Changed

- `register_commands.py` - Added `/location` command definition
- `src/handlers/location_handler.py` - New modular handler module
- `src/handlers/discord_handler.py` - Added location command routing

## How to Test

The Lambda function is already deployed and running in AWS at: [trainee-2026-najmul-mhp-discord-bot](https://ap-southeast-1.console.aws.amazon.com/lambda/home?region=ap-southeast-1#/functions/trainee-2026-najmul-mhp-discord-bot)

1. Register commands with Discord:
   - `python register_commands.py`
2. Join the test server and run commands in #1-location-cmd-test: https://discord.gg/f5yhCxvP
3. If commands produce no reply, check CloudWatch:
   - `aws logs tail /aws/lambda/trainee-2026-najmul-mhp-discord-bot --follow`

## How QA Should Test

**Command Test Cases:**

| Command | Expected Reply | Notes |
|---------|---|---|
| `/location location: office date: today` | ✅ **Office** set for 2026-03-13 🏢 Location: OFFICE | Valid today date |
| `/location location: wfh date: 2026-03-15` | ✅ **Work From Home** set for 2026-03-15 🏠 Location: WFH | Valid future date (within 7 days) |
| `/location location: office date: 2026-03-20` | ⚠️ Cannot update location more than 7 days in advance. Maximum date: 2026-03-20 | Date beyond 7-day window |
| `/location location: office date: 2026-03-10` | ⚠️ Cannot update location for past dates. | Past date (before today) |
| `/location location: office date: invalid` | ⚠️ Invalid date format. Use "today" or YYYY-MM-DD (e.g., 2026-03-15) | Malformed date |
| `/location location: office date: today` (after 9 PM previous day) | ⚠️ Location cut-off for 2026-03-13 was 2026-03-12 9:00 PM (Asia/Dhaka). Updates are no longer allowed. | Cut-off enforced |
| `/location location: office date: today` (user doesn't exist) | ⚠️ User not found. Please use /meal command first to create your profile. | No auto-creation |

**Success Criteria:**
- ✅ Valid location updates return confirmation message
- ✅ Invalid dates rejected with clear error messages
- ✅ Past dates blocked
- ✅ Out-of-window dates (>7 days) blocked
- ✅ Cut-off rule enforced at D-1 9:00 PM (Asia/Dhaka)
- ✅ User existence validated
- ✅ All replies are ephemeral
- ✅ DynamoDB stores location with team_id for filtering

## Rollback Plan

If this PR causes issues in production:
1. **Immediate action:** Revert this PR to remove the `/location` command
2. **No data migration needed** - Location entries are optional;
3. **Safe to revert** - No breaking changes to existing meal command or user profiles
4. **Verification:** After revert, confirm `/location` no longer appears in Discord bot command list

## Review Focus

- correctness of cut-off logic and timezone handling (Asia/Dhaka)
- date validation branches and edge cases
- DynamoDB persistence with team denormalization
- user existence validation (no auto-creation behavior)
- modular handler structure and code organization
- response message clarity and format

## Checklist

- [X] `/location` command registered and visible in Discord
- [X] `/location` handler wired in command router
- [X] Location handler extracted to separate module (`location_handler.py`)
- [X] Date parsing/validation implemented (`today` + `YYYY-MM-DD`)
- [X] 7-day forward window enforced
- [X] Previous-day 9 PM cut-off enforced (`Asia/Dhaka`)
- [X] DynamoDB persistence integrated with team denormalization
- [X] Ephemeral error handling in place
- [X] Response message simplified
- [X] Code is modular and maintainable

## Note for Reviewer

This PR introduces modular handler architecture by extracting location logic to a separate module. This pattern should be replicated for future commands (e.g., `/mystatus`, `/team-summary`) to maintain code organization and readability as the bot grows.

### CODE REVIEW

**Review by rifat-craftsmen — CHANGES_REQUESTED** [review]

**Review by mehedi-1101 — APPROVED** [review]
No blockers.lgtm

File: src/handlers/location_handler.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        # Extract command options
+        data = interaction.get('data', {})
+        options = {opt['name']: opt['value'] for opt in data.get('options', [])}
+        
+        location_str = options.get('location')
+        date_str = options.get('date')  # Required, can be 'today' or YYYY-MM-DD
+        
+        # Extract user info
+        member = interaction.get('member', {})
+        user_data = member.get('user', {})
```
[the remark this answers is not in the document]
Comment (najmul-craftsmen) [reply src/handlers/location_handler.py]: fixed [824e784](https://github.com/najmul-craftsmen/Project-MHP/pull/20/changes/824e78437d22d2048912c120b5d5920804f18e7f)

**Review by rifat-craftsmen — APPROVED** [review]
